common/persistence/health_signal_aggregator.go

```diff
@@ -0,0 +1,172 @@
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package persistence
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/aggregate"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/metrics"
+	"go.temporal.io/server/common/quotas"
+)
+
+const (
+	emitMetricsInterval = 30 * time.Second
+)
+
+type (
+	HealthSignalAggregator interface {
+		common.Daemon
+		GetRecordFn(req quotas.Request) func(err error)
+		AverageLatency() float64
+		ErrorRatio() float64
+	}
+
+	HealthSignalAggregatorImpl struct {
+		status     int32
+		shutdownCh chan struct{}
+
+		requestsPerShard map[int32]int64
+		requestsLock     sync.Mutex
+
+		latencyAverage aggregate.MovingWindowAverage
+		errorRatio     aggregate.MovingWindowAverage
+
+		metricsHandler       metrics.Handler
+		emitMetricsTimer     *time.Ticker
+		perShardRPSWarnLimit dynamicconfig.IntPropertyFn
+
+		logger log.Logger
+	}
+)
+
+func NewHealthSignalAggregatorImpl(
+	windowSize dynamicconfig.DurationPropertyFn,
+	maxBufferSize dynamicconfig.IntPropertyFn,
+	metricsHandler metrics.Handler,
+	perShardRPSWarnLimit dynamicconfig.IntPropertyFn,
+	logger log.Logger,
+) *HealthSignalAggregatorImpl {
+	return &HealthSignalAggregatorImpl{
+		status:               common.DaemonStatusInitialized,
+		shutdownCh:           make(chan struct{}),
+		requestsPerShard:     make(map[int32]int64),
+		latencyAverage:       aggregate.NewMovingWindowAvgImpl(windowSize(), maxBufferSize()),
+		errorRatio:           aggregate.NewMovingWindowAvgImpl(windowSize(), maxBufferSize()),
+		metricsHandler:       metricsHandler,
+		emitMetricsTimer:     time.NewTicker(emitMetricsInterval),
+		perShardRPSWarnLimit: perShardRPSWarnLimit,
+		logger:               logger,
+	}
+}
+
+func (s *HealthSignalAggregatorImpl) Start() {
+	if !atomic.CompareAndSwapInt32(&s.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+	go s.emitMetricsLoop()
+}
+
+func (s *HealthSignalAggregatorImpl) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+	close(s.shutdownCh)
+	s.emitMetricsTimer.Stop()
+}
+
+func (s *HealthSignalAggregatorImpl) GetRecordFn(req quotas.Request) func(err error) {
+	start := time.Now()
+	return func(err error) {
+		s.latencyAverage.Record(time.Since(start).Milliseconds())
+
+		if isUnhealthyError(err) {
+			s.errorRatio.Record(1)
+		} else {
+			s.errorRatio.Record(0)
+		}
```

Comment:
I would comment them out since they are not actually being used right now. We can uncomment them probably in the dynamic rate limiter PR.